pi/imu.py

```diff
@@ -1,59 +1,89 @@
 PORT = 2049  # The port used by the server
-CONTROL_LOOP_FREQ = 5  # Hz
+CONTROL_LOOP_FREQ = 100  # Hz
+
+def read_data() -> dict:
+    data = {}
+    # acceleration
+    accel_x, accel_y, accel_z = bno.acceleration  # pylint:disable=no-member
+    acceleration = {}
+    acceleration["x"] = accel_x
+    acceleration["y"] = accel_y
+    acceleration["z"] = accel_z
+    data["acceleration"] = acceleration
 
-# read acceleration and rotation
-def read_data(sensor):
-        data_dict = {}
-        accel_x, accel_y, accel_z = sensor.linear_acceleration  # pylint:disable=no-member
-        data_dict["acceleration"] = ("X: %0.6f  Y: %0.6f Z: %0.6f  m/s^2" % (accel_x, accel_y, accel_z))
+    # gyro
+    gyro_x, gyro_y, gyro_z = bno.gyro  # pylint:disable=no-member
+    gyroscope = {}
+    gyroscope["x"] = gyro_x
+    gyroscope["y"] = gyro_y
+    gyroscope["z"] = gyro_z
+    data["gyroscope"] = gyroscope
 
-        quat_i, quat_j, quat_k, quat_real = sensor.game_quaternion  # pylint:disable=no-member
-        data_dict["rotation vector"] = (
-            "I: %0.6f  J: %0.6f K: %0.6f  Real: %0.6f" % (quat_i, quat_j, quat_k, quat_real)
-        )
+    # magnetometer
+    mag_x, mag_y, mag_z = bno.magnetic  # pylint:disable=no-member
+    magnetometer = {}
+    magnetometer["x"] = mag_x
+    magnetometer["y"] = mag_y
+    magnetometer["z"] = mag_z
+    data["magnetometer"] = magnetometer
 
-        return data_dict
+    # quaternion
+    quat_i, quat_j, quat_k, quat_real = bno.quaternion  # pylint:disable=no-member
+    quaternion = {}
+    quaternion["i"] = quat_i
+    quaternion["j"] = quat_j
+    quaternion["k"] = quat_k
+    quaternion["real"] = quat_real
+    data["quaternion"] = quaternion
+
+    return data
 
 with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
     print(f"connecting to {HOST}:{PORT}")
     s.connect((HOST, PORT))
     last_time = time.time()
-    # initialize sensor
-    i2c = busio.I2C(board.SCL, board.SDA, frequency=400000)
-    bno = BNO08X_I2C(i2c)
-
-    # leave these in in case we switch to uart:
-    # uart = serial.Serial("/dev/serial0", 115200)
-    # uart = busio.UART(board.TX, board.RX, frequency=3000000, receiver_buffer_size=2048)
-
-    bno.enable_feature(BNO_REPORT_GAME_ROTATION_VECTOR)
-    bno.enable_feature(BNO_REPORT_LINEAR_ACCELERATION)
-
+    
     while True:
-        data = read_data(bno)
+        data = read_data()
+        msg = {
+            "imu_data": json.dumps(data),
+        }
+        s.send(str.encode(json.dumps(msg)))
 
-        # send to async server as json
-        msg = json.dumps(data)
-        s.send(str.encode(msg))
         print(f"sent: {msg}")
 
         # sleep for remainder of loop
         if time.time() - last_time < 1 / CONTROL_LOOP_FREQ:
             time.sleep(1 / CONTROL_LOOP_FREQ - (time.time() - last_time))
         else:
             print("Warning: control loop took too long")
-        last_time = time.time()
-
+        last_time = time.time()
```

Comment:
Generally would want a little more documentation so that if someone who has never seen this before can understand what is going on

Reply:
Sitting next to Jonah here while he reviews and want to echo some comments about control frequency and abstracting some of that instantiation